Cleanup host veth side when a pod is gone

phoracek opened the pull request:
With this patch, we assign pod side of veth pair not to pod netns directly, byt to a named netns linking to the pod's netns. By doing that, removal of pod, removes also host side of veth pair.

All we need to do is to clean up dead named netns symlinks.

Please, review only the second commit.

phoracek committed:
add functional test for bridge dp

With this test we try to deploy the plugin, attach Pods to provided
resources and ping between them.

phoracek commented:
In CNI bridge plugin, they create the pair inside container netns and then move one side to host, that may be even better solution without a need for named netns.

yuvalif commented on pkg/network/bridge/plugin.go:
```diff
@@ -274,8 +284,32 @@ func (nbdp *NetworkBridgeDevicePlugin) attachPodToBridge(bridgeName, nicName str
+	netnsName := fmt.Sprintf("%s%d", namespacePrefix, rand.Intn(999999999))
+
+	// ensure that /var/run/netns directory will be initialized
+	exec.Command("ip", "netns", "add", "dummyNS").Run()
```
why not use the netns library?

yuvalif commented:
this is also what @rmohr  suggest here: https://github.com/kubevirt/kubevirt/pull/1407/files/4601c807d248223522166f336d218fffebfc0d6c..202946747ee3fe810a3a3724fcabc19d01a8f6f8#diff-5b80ff518958e93de348b14803f57c34
seems like the best option for cleanup.

phoracek committed:
bridge: clean up both sides of veth

Create veth pair inside Pod netns, so both sides will disappear once
the Pod is gone.

phoracek commented:
@yuvalif @rmohr I changed it, so it creates the veth inside pod netns, but it seems like it did not help, `nic_` stays on host until the DP is removed.

rmohr commented:
@phoracek are you creating int in the namespace of the Pod of the DP itself or in the target Pod?

phoracek commented:
@rmohr I use container PID to enter the namespace, then being there, I create veth pair and move veth peer to original (DP) namespace.

https://github.com/kubevirt/kubernetes-device-plugins/pull/50/commits/59c78e3c70df1f07a05a7aad03e7f5994561ec50#diff-4f9899f5b818b28f11bb7bebaaf54390R280

yuvalif commented on pkg/network/bridge/plugin.go:
```diff
@@ -237,107 +236,138 @@ func (nbdp *NetworkBridgeDevicePlugin) attachPods() {
+
+func inNetnsDo(targetNs netns.NsHandle, callback func(originalNs netns.NsHandle) error) error {
+	// store current namespace
+	originalNs, err := netns.Get()
```
maybe you need to create a specific namespace to be used by the DS pod on the host side, and not just use whatever you get from netns.Get()?this is the only difference i see from the other examples.

phoracek replied:
There are two other differences.

One is that in other solutions, one side of veth remains in a namespace bound to a temporary one-shot process (CNI plugin), we use long running daemon. That would explain why we remove it only when the DS goes down.

Another difference might be that in some examples, they use named namespaces (in /var/run/netns/).

rmohr replied:
So `virt-handler` is also long-running and it works in this PR https://github.com/kubevirt/kubevirt/pull/1407 flawless. There should also no connection to a process. 

Very unlikely, but did you check if your Pods are stuck in "terminating"? Maybe the pods do not get cleaned up?

phoracek replied:
@rmohr  It took them quite a long time to die, I will check, it sounds likely.

phoracek replied:
They are quite long in "terminating" but they end up removed, still not cleaned up.

yuvalif replied:
@phoracek can you try to create a namespace for the host side as well?

phoracek replied:
I will. I just rewrote it using CNI libraries, still not cleaning up.

phoracek replied:
Finally with CNI libraries used, locking and Go 1.10 it works. Tests are raceful, but when tested manually, it works.

phoracek commented:
Tests are passing now, cleanup works. Will finish this after #47
